torch/utils/checkpoint.py

```diff
@@ -159,31 +162,31 @@ def add_device_types(arg):
 # the device of all Tensor args.
 #
 # To consider:  maybe get_device_states and set_device_states should reside in torch/random.py?
-def get_device_states(*args) -> Tuple[List[int], List[torch.Tensor]]:
+def get_device_states(*args) -> Tuple[List[torch.device], List[torch.Tensor]]:
     # This will not error out if "arg" is a CPU tensor or a non-tensor type because
     # the conditionals short-circuit.
-    fwd_device_ids = []
+    fwd_devices = []
 
-    def add_device_ids(arg):
-        nonlocal fwd_device_ids
+    def add_device(arg):
+        nonlocal fwd_devices
         if isinstance(arg, torch.Tensor) and not arg.device.type == "cpu":
-            fwd_device_ids.append(arg.get_device())
-    tree_map(add_device_ids, args)
+            fwd_devices.append(arg.device)
+    tree_map(add_device, args)
 
     fwd_device_states = []
     device_module = _get_device_module(_infer_device_type(*args))
 
-    for device_id in fwd_device_ids:
-        with device_module.device(device_id):
+    for device in fwd_devices:
+        with device_module.device(device):
             fwd_device_states.append(device_module.get_rng_state())
 
-    return fwd_device_ids, fwd_device_states
+    return fwd_devices, fwd_device_states
 
 
 def set_device_states(devices, states) -> None:
```

Comment:
Could we have set_device_states accept a "device" argument?
Then for both callsites (`use_reentrant={True,False}`) it is easy to access the device from backward, for `True` (custom autograd Function) path, you can use `ctx.device`, and for the `False` (saved tensor hook) path, you can use the closed over "device" variable.

Reply:
Yes. Done.